src/megaapi_impl.cpp

```diff
@@ -882,6 +930,194 @@ MegaHandle MegaNodePrivate::getOwner() const
 
 const unsigned int MegaApiImpl::MAX_SESSION_LENGTH = 64;
 
+
+MegaBackgroundMediaUploadPrivate::MegaBackgroundMediaUploadPrivate(MegaApiImpl* capi)
+    : api(capi)
+{
+    // generate fresh random encryption key/CTR IV for this file
+    api->client->rng.genblock(filekey, sizeof filekey);
+}
+
+MegaBackgroundMediaUploadPrivate::MegaBackgroundMediaUploadPrivate(const string& serialised, MegaApiImpl* capi)
+    : api(capi)
+{
+    CacheableReader r(serialised);
+    string mediapropertiesstr;
+    unsigned char expansions[8];
+    string fileattrstring; // fileattrstring is not serialized
```

Comment:
I know Simon is keen for that - initially we had too much else to get working and using the existing functions for attaching those to an existing node kept it a bit simpler, with less to go wrong.   I think we could do it by providing a version of those functions like setThumbnail but without taking a Node, and just do the metad uploading (using the existing framework) and return a file attribute to the listener.  Those file attributes could then be supplied to the backgroundMediaUploadComplete() for the "atomic" completion that Simon is looking to achieve- where the concern is that once the user sees the node appear in the browser, they may close or delete the iOS app, without all attributes or file attributes being attached yet.

Reply:
True. It's a tricky use-case. From other point of view, we could say the _media-upload_ does not finishes until its metadata are also uploaded and attached to the node.
I assume that Simon, in case he's working on that improvement, it's going to be done in a different branch than the main feature's branch. Hence, we can leave the required changes in the SDK for another branch too. But please, if so, create a ticket with the introduction to the issue, the link to the expected behavior explained by Mathias, and the suggested workaround.
Thanks

Reply:
Hi Sergio, good point.  We're so close to completing CU2 (the SDK part anyway), I thought it would be great to finalize that, so the interface is stable when it gets to `develop`.   I've pushed a change so the thumbnail and preview can be pushed with the initial putnodes, please take a look when you get a chance.  thanks!